File: targets/widget/SharedStore.swift
import Foundation

/// One server as mirrored from the React Native app into the shared App Group.
/// Written by `utils/widgetSync.ts` via `ExtensionStorage.set("servers", JSON.stringify(...))`.
struct StoredServer: Codable, Identifiable {
  let id: String
  let title: String?
  let url: String
  let username: String?
  let password: String?
  let authRequired: Bool?

  var displayTitle: String {
    if let t = title, !t.trimmingCharacters(in: .whitespaces).isEmpty { return t }
    return url
  }
}

/// Reads the server list the app writes into the shared App Group container.
enum SharedStore {
  static let appGroup = "group.io.evcc.app"

  private static var defaults: UserDefaults? { UserDefaults(suiteName: appGroup) }

  /// `ExtensionStorage.set(key, string)` stores a plain string, so read it back
  /// as a string and decode the JSON ourselves.
  static func servers() -> [StoredServer] {
    guard let json = defaults?.string(forKey: "servers"),
          let data = json.data(using: .utf8),
          let list = try? JSONDecoder().decode([StoredServer].self, from: data)
    else { return [] }
    return list
  }

  static func activeServerId() -> String? {
    defaults?.string(forKey: "activeServerId")
  }

  static func defaultServer() -> StoredServer? {
    let all = servers()
    if let id = activeServerId(), let match = all.first(where: { $0.id == id }) {
      return match
    }
    return all.first
  }

  /// Resolve the server selected in the widget config, falling back to the active one.
  static func server(id: String?) -> StoredServer? {
    guard let id, !id.isEmpty else { return defaultServer() }
    return servers().first(where: { $0.id == id }) ?? defaultServer()
  }

  /// Timestamp of the last mode change made from the widget, so the loadpoint
  /// provider can schedule quick follow-up reloads (vehicles react slowly).
  static func markModeChanged() {
    defaults?.set(Date().timeIntervalSince1970, forKey: "lastModeChange")
  }

  static func secondsSinceModeChange() -> TimeInterval? {
    let t = defaults?.double(forKey: "lastModeChange") ?? 0
    return t > 0 ? Date().timeIntervalSince1970 - t : nil
  }
}

File: targets/android-widget/kotlin/SharedStore.kt
package io.evcc.android.widget

import android.content.Context
import org.json.JSONObject
import java.io.File

/** One server, mirrored from the React Native app (utils/widgetSync.ts). */
data class StoredServer(
    val id: String,
    val title: String?,
    val url: String,
    val username: String?,
    val password: String?,
    val authRequired: Boolean,
) {
    val displayTitle: String
        get() = title?.takeIf { it.isNotBlank() } ?: url
}

/**
 * Reads the server list the RN app writes to the app's document directory.
 * Android counterpart of the iOS SharedStore.swift (App Group). Same package
 * as the app, so no App Group / native module is needed — a plain file works.
 */
object SharedStore {
    // keep in sync with widgetSync.ts ANDROID_SERVERS_FILE
    private const val FILE_NAME = "evcc-widget-servers.json"

    private fun readJson(context: Context): JSONObject? {
        val f = File(context.filesDir, FILE_NAME)
        if (!f.exists()) return null
        return runCatching { JSONObject(f.readText()) }.getOrNull()
    }

    fun servers(context: Context): List<StoredServer> {
        val arr = readJson(context)?.optJSONArray("servers") ?: return emptyList()
        return (0 until arr.length()).mapNotNull { i ->
            val o = arr.optJSONObject(i) ?: return@mapNotNull null
            StoredServer(
                id = o.optString("id"),
                title = o.optString("title").takeIf { it.isNotEmpty() },
                url = o.optString("url"),
                username = o.optString("username").takeIf { it.isNotEmpty() },
                password = o.optString("password").takeIf { it.isNotEmpty() },
                authRequired = o.optBoolean("authRequired", false),
            )
        }
    }

    private fun activeServerId(context: Context): String? =
        readJson(context)?.optString("activeServerId")?.takeIf { it.isNotEmpty() }

    fun defaultServer(context: Context): StoredServer? {
        val all = servers(context)
        val id = activeServerId(context)
        return all.firstOrNull { it.id == id } ?: all.firstOrNull()
    }

    /** Resolve the server selected in the widget config, else the active one. */
    fun server(context: Context, id: String?): StoredServer? {
        if (id.isNullOrEmpty()) return defaultServer(context)
        return servers(context).firstOrNull { it.id == id } ?: defaultServer(context)
    }
}
